src/wallet/wallet.cpp

```diff
@@ -2334,12 +2334,22 @@ DBErrors CWallet::LoadWallet()
 util::Result<void> CWallet::RemoveTxs(std::vector<uint256>& txs_to_remove)
 {
     AssertLockHeld(cs_wallet);
-    WalletBatch batch(GetDatabase());
-    if (!batch.TxnBegin()) return util::Error{_("Error starting db txn for wallet transactions removal")};
+    util::Result<std::vector<CWallet::MapWalletIt>> res{util::Error{}}; // Note: Make RunWithinTxn generic.
+    bool was_txn_committed = RunWithinTxn(GetDatabase(), /*process_desc=*/"remove transactions", [&](WalletBatch& batch) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
+        return (res = RemoveTxs(batch, txs_to_remove)).has_value();
+    });
+    if (!res) return util::Error{util::ErrorString(res)};
+    if (!was_txn_committed) return util::Error{_("Error starting/committing db txn for wallet transactions removal process")};
 
+    // Now that txs have been deleted from disk, erase them from memory.
+    ClearTxs(*res);
+    return {};
+}
+
+util::Result<std::vector<CWallet::MapWalletIt>> CWallet::RemoveTxs(WalletBatch& batch, std::vector<uint256>& txs_to_remove) {
```

Comment:
In 4b3ec698a228fc531cca946f9a572e1dbdc4b0d5 "wallet: prepare RemoveTxs to be used within a broader db txn procedure"

I don't really like this refactor. This means that any callers of this particular overload will have to remember to do `ClearTxs`, whereas callers of the other one don't. This seems like something that could result in programming mistakes in the future. Generally our overloads should do basically the same things, just with the option of providing different arguments. Not clearing the txs from `mapWallet` in one overload is a pretty big difference that may not be easily noticed.

I think it would actually be a bit cleaner and better to have `RemoveTxs` detect if a database transaction has been started already and just choose to not begin a transaction if that's the case.

Reply:
Ended up introducing a mechanism to listen for the db txn outcomes. So internal procedures can register callbacks to update the in-memory state only when the data is successfully committed to disk.